content/academy/puppeteer_playwright/executing_scripts.md

```diff
@@ -0,0 +1,143 @@
+
+An important concept to understand when dealing with headless browsers is the **context** in which your code is being run. For example, if you try to use the native `fs` Node.js module (used in the previous lesson) while running code in the context of the browser, errors will be thrown saying that it is undefined. Similarly, if you are trying to use `document.querySelector()` or other browser-specific functions in the server-side Node.js context, errors will also be thrown.
+
+Here is an example of a common mistake made by beginners to Puppeteer/Playwright:
+
+```JavaScript
+// THIS IS INCORRECT CODE
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+// visit google
+await page.goto('https://google.com/');
+
+// change background to green
+document.body.style.background = 'green';
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+```
+
+When we try and run this, we get this error:
+
+```text
+ReferenceError: document is not defined
+```
+
+The reason this is happening is because we're trying to run browser-side code on the server-side where it is not supported. [`document`](https://developer.mozilla.org/en-US/docs/Web/API/Document) is a property of the browser [**Window**](https://developer.mozilla.org/en-US/docs/Web/API/Window) instance that holds the rendered website; therefore, this API is not available in Node.js. So how are we supposed to run code within the context of the browser?
+
+## [](#running-in-browser-context) Running code in the context of the browser
+
+We will use `page.evaluate()` to run our code in the browser. This method takes a callback as its first parameter, which will be executed within the browser.
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+await page.evaluate(() => {
+    document.body.style.background = 'green';
+});
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+await page.evaluate(() => {
+    document.body.style.background = 'green';
+});
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+```
+
+Here's what we see in the automated browser when we run this code:
+
+![Google with the background color changed to green]({{@asset puppeteer_playwright/images/green-google.webp}})
+
+## [](#using-variables-in-page-evaluate) Using variables in `page.evaluate()`
+
+Within our code, we generate a `randomString` in the Node.js context:
+
+```JavaScript
+const randomString = Math.random().toString(36).slice(2);
+```
```

Comment:
How about just passing a different color instead of the random string? To simplify it and connect it with the previous example. Say something like "but what if the information about the color we want is only available in the Node.js context? We can pass a variable...."

Reply:
I considered doing that, but since it's our first time showing that we can execute scripts, I really wanted to show them a few examples of what they can do. Change the background color, change the title - now they understand that they can do pretty much anything.

Reply:
Yeah, my point was mostly around the random number generation. We can change the title to `Apify is Awesome` and it will work the same without adding the complexity of generating a random string.